cmd/kured/main.go

```diff
@@ -336,26 +436,44 @@ func rebootAsRequired(nodeID string, window *timewindow.TimeWindow, TTL time.Dur
 		release(lock)
 	}
 
+	if !rebootRequired() {
+		preferNoSchedule(client, nodeID, false)
+	}
+
 	source := rand.NewSource(time.Now().UnixNano())
 	tick := delaytick.New(source, period)
 	for range tick {
-		if window.Contains(time.Now()) && rebootRequired() && !rebootBlocked(client, nodeID) {
-			node, err := client.CoreV1().Nodes().Get(context.TODO(), nodeID, metav1.GetOptions{})
-			if err != nil {
-				log.Fatal(err)
-			}
-			nodeMeta.Unschedulable = node.Spec.Unschedulable
-
-			if acquire(lock, &nodeMeta, TTL) {
-				if !nodeMeta.Unschedulable {
-					drain(client, node)
-				}
-				commandReboot(nodeID)
-				for {
-					log.Infof("Waiting for reboot")
-					time.Sleep(time.Minute)
-				}
-			}
+		if !window.Contains(time.Now()) {
+			continue
+		}
+
+		if !rebootRequired() {
+			continue
+		}
+
+		preferNoSchedule(client, nodeID, true)
+
+		if rebootBlocked(client, nodeID) {
+			continue
+		}
+
+		node, err := client.CoreV1().Nodes().Get(context.TODO(), nodeID, metav1.GetOptions{})
+		if err != nil {
+			log.Fatal(err)
+		}
+		nodeMeta.Unschedulable = node.Spec.Unschedulable
+
+		if !acquire(lock, &nodeMeta, TTL) {
+			continue
```

Comment:
I am confused as how preferNoSchedule works. 

I thought that the preferNoSchedule flag would be set (3rd argument to true) when:
- The rebootRequired returns true
- The node cannot get a lock because another node has a lock.
Therefore I expected the line 454 to be in here (L467) instead, before the continue. However, if we agree that all nodes should be marked as NoSchedule immediately when we know they need rebooting (during the maintenance window), then it's fine for me. After all, the code is indeed cleaner in the current PR implementation.
As long as we document how we do it (maybe a little comment on L454 would help?).

However, I expected the function to flip the NoSchedule to false at all ticks, not simply at the restart of the daemonset. If the timewindow to trigger all the reboots is too small, some nodes might not have rebooted in the maintenance window, and therefore some ndoes will still marked with no schedule outside the timewindow. That doesn't seem right.

TL;DR: Therefore I would move the L440 inside the block 450-452, move that block above the block of the timewindow (446), move the L454 `preferNoSchedule` in the block L466-468 before the `continue`. But of course, I am easy, and happy to discuss it here.

Reply:
i assumed the order of the checks would be like this because the coast to execute them.
the window-check is just a calculation, the requires-reboot is a local check and the reboot-blocked check is a rpc.

i did not want to have the remove-taint within the loop, because it could take a long time until a rebooted node get untained.
maybe a additional untaint in the block from 450-453?

The true-false flag was a pain in my eye anyway. What do you think about two functions addPreferNoScheduleTaint and removePreferNoScheduleTaint?

Reply:
> This PR changes the state of the taint `TaintEffectPreferNoSchedule` regardless of not it was tained by kured. This is a problem for me, because one administrator might want to mark a node for NoSchedule for other reasons, and we might remove it without its approval. WDYT?

Only the taint "weave.works/kured-node-reboot" (by default) is added or remove by the new behavior. Other taints will be untouched. The code from https://github.com/damoon/kured/blob/19-PreferNoSchedule/cmd/kured/main.go#L326 until https://github.com/damoon/kured/blob/19-PreferNoSchedule/cmd/kured/main.go#L367 ensures only to add/remove this one taint.

Reply:
For addPreferNoScheduleTaint: That's indeed more explicit (especially in this part of the code). But overall it's up to you, you wrote it. It's no big deal for me. However: Let's not copy the function into two (and repeat ourselves):  that sounds even worse :)

I am not sure it's about the cost to execute. Someone with more history of the project shoudl tell :)

I was not fond of having this additional untaint without explanations. For me, all this needs is comments, to explain what the design decisions we made. For example:
`we do this outside the main loop because it's an API task, so it takes more time, for which we want to avoid running in the tick loop. This means that a temporary API issue will cause a crash of the daemonset until its state is correct. This also means that if the maintenance window is too small, the machines can still technically be marked with no schedule outside out the maintenance window`.

Reply:
I added a comment for the untaint outside the loop,
moved the tainting inside the check for rebootBlocked,
and added a untaint to allow for the the reboot window.

Taints got their own package and better naming, the bool is hidden.

Does this look good to you?

Reply:
:+1: